perf/impl/https/main.go

```diff
@@ -0,0 +1,208 @@
+
+	for bytesToSend > 0 {
+		toSend := buf
+		if bytesToSend < BlockSize {
+			toSend = buf[:bytesToSend]
+		}
+
+		n, err := w.Write(toSend)
+		if err != nil {
+			http.Error(w, "Failed write", http.StatusInternalServerError)
+			return
+		}
+		bytesToSend -= uint64(n)
+	}
+}
+
+func runClient(serverAddr string, uploadBytes, downloadBytes, nTimes int) ([]time.Duration, error) {
+	durations := make([]time.Duration, nTimes)
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+
+	reqBody := make([]byte, 8+uploadBytes)
+	binary.BigEndian.PutUint64(reqBody, uint64(downloadBytes))
+
+	for i := 0; i < nTimes; i++ {
+		startTime := time.Now()
+		resp, err := client.Post(fmt.Sprintf("https://%s/", serverAddr), "application/octet-stream", bytes.NewReader(reqBody))
+		if err != nil {
+			return durations, err
+		}
+
+		respBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			fmt.Printf("Error reading response: %v\n", err)
+			return durations, err
+		} else if len(respBody) != downloadBytes {
+			fmt.Printf("Expected %d bytes in response, but received %d\n", downloadBytes, len(respBody))
+			return durations, err
+		}
+		resp.Body.Close()
+
+		durations[i] = time.Since(startTime)
+	}
+
+	return durations, nil
+}
+
+func generateEphemeralCertificate() (tls.Certificate, error) {
+	// Generate a private key
+	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+
+	// Set up the certificate template
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			Organization: []string{"Ephemeral Cert"},
+		},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().Add(24 * time.Hour),
+		KeyUsage:  x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageServerAuth,
+		},
+		BasicConstraintsValid: true,
+	}
+
+	// Set the IP address if required
+	ip := net.ParseIP("127.0.0.1")
+	if ip != nil {
+		template.IPAddresses = append(template.IPAddresses, ip)
+	}
+
+	// Create a self-signed certificate
+	certDER, err := x509.CreateCertificate(rand.Reader, &template, &template, &privKey.PublicKey, privKey)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+
+	// PEM encode the certificate and private key
+	certPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certDER})
+	privKeyPEM := pem.EncodeToMemory(&pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(privKey)})
+
+	// Create a tls.Certificate from the PEM encoded certificate and private key
+	cert, err := tls.X509KeyPair(certPEM, privKeyPEM)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+
+	return cert, nil
+}
+
+type Latencies struct {
+	Latencies []float32 `json:"latencies"`
+}
+
+func main() {
+	runServer := flag.Bool("run-server", false, "Should run as server")
+	serverIPAddr := flag.String("server-ip-address", "", "Server address")
```

Comment:
Can we not hard-code the port? `IP:port` would be an easy way to be more general.

Reply:
What do you need different ports for @marten-seemann?

Reply:
I don’t see any reason to hardcode this. I’ve already run into the problem that I want to use a different port (since this is the Kubo default port), and this seems super trivial to make configurable.

Reply:
> and this seems super trivial to make configurable.

Not super trivial. Thus far the *Runner* does not have to take any information from the *Server* to the *Client*, e.g. by reading the *Server's* stdout. In case we want to make the port dynamic at runtime, we would need some mechanism from server->runner-> client. None of this is very difficult. Though I will not prioritize it e.g. over proper visualizations unless we have a direct use-case for it.

Reply:
Added to the outstanding list in the pull request description.

Reply:
I think we have a misunderstanding here. I wasn't asking for the server picking a random port, and communicating that to the client. I don't think we need that at all.

Instead, I was expecting that the runner passes the port to the server:
```
go run main.go -run-server=true -server-address=0.0.0.0:1234
```

fwiw, that's how I implemented it in quic-go's perf runner: https://github.com/quic-go/perf/blob/master/cmd/main.go

Does that make sense?